### PR TITLE
feat(web): add a text chat using the existing connection

### LINKED ISSUE #28: Create a simple chat
Sometimes is important write something to solve a miss understanding of something, or to share some link like twitter/linkedin/github.

So will be good if we have a chat using WebRTC.

### PR DESCRIPTION
**Pull request by joaoreider**
Closes #28
 
## Demo
[chat-demo.webm](https://github.com/brunocroh/bolhadev.chat/assets/83885051/012a2a25-7cd9-42f0-a340-21c26d165937)

### CODE REVIEW

File: apps/web/components/chat.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import { ScrollArea } from './ui/scroll-area'
+
+interface ChatProps {
+  messages: Message[]
+  onSend: (message: string) => void
+}
+
+export function Chat({ messages, onSend }: ChatProps) {
+  const [inputValue, setInputValue] = useState('')
+  const inputRef = useRef<HTMLInputElement>(null)
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]: inputRef is unecesseary here, please remove it

File: apps/web/components/chat.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  const sendMessage = () => {
+    if (inputValue.trim() !== '') {
+      const message = inputValue.trim()
+      onSend(message)
+      setInputValue('')
+    }
+  }
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]:
```suggestion
  const sendMessage = () => {
  const message = inputValue.trim()
    if (message) {
      onSend(message)
      setInputValue('')
    }
  }
```

File: apps/web/components/chat.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setInputValue(event.target.value)
+  }
+
+  return (
+    <div className="flex w-full flex-col items-center gap-2">
+      <ScrollArea className="flex h-48 w-1/2 flex-col justify-center gap-2 rounded-lg border p-4">
+        {messages.length === 0 ? (
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]:
```suggestion
        {!messages.length ? (
```

File: apps/web/components/chat.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          onKeyDown={handleKeyPress}
+          onChange={handleInputChange}
+        />
+        <Button onClick={sendMessage}>Send</Button>
+      </div>
+    </div>
+  )
+}
+
+const MessageItem = ({ message }: { message: Message }) => {
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]: use React.FC to type components

File: apps/web/components/chat.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  const isMe = message.sender === 'Me'
+  return (
+    <div
+      className={`flex items-center gap-3 justify-${isMe ? 'start' : 'end'}`}
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]: use clsx to conditionally add types

```suggestion
      className={clsx('flex items-center gap-3', isMe && 'justify-start' || 'justify-end)}
```

File: apps/web/components/chat.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    >
+      <span
+        className={`bg-${isMe ? 'accent' : 'purple-950'} m-1 max-w-xs rounded-md p-3`}
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]: use clsx

File: apps/web/components/chat.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      >
+        <span className="text-sm font-semibold">{message.sender}</span>:{' '}
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]: use Shadcn badge with different colors to be clear to user what message is from him or other user.

File: apps/web/app/room/[slug]/page.tsx [hunk 8]
```diff
@@ -11,12 +11,18 @@ import useWebSocket from 'react-use-websocket'
...
 import { Button } from '@/components/ui/button'
 import { Card, CardContent } from '@/components/ui/card'
 import { VideoPlayer } from '@/components/video-player'
 import { useUserMedia } from '@/hooks/useUserMedia'
 
+export type Message = {
+  sender: 'Me' | 'Other'
+  content: string
+}
+
```
Comment (brunocroh) [inline apps/web/app/room/[slug]/page.tsx]: move this type to Chat component

File: apps/web/app/room/[slug]/page.tsx [hunk 9]
```diff
@@ -103,6 +110,13 @@ export default function Page(): JSX.Element {
               }
             })
 
+            peerRef.current?.on('data', (message) => {
+              setMessages((prevMessages) => [
+                ...prevMessages,
+                { sender: 'Other', content: JSON.parse(message.toString()) },
```
Comment (brunocroh) [inline apps/web/app/room/[slug]/page.tsx]: Check if is really necessary `JSON.parse` here

File: apps/web/app/room/[slug]/page.tsx [hunk 10]
```diff
@@ -192,6 +206,16 @@ export default function Page(): JSX.Element {
     location.replace(`/room/queue`)
   }, [stopAllStreaming])
 
+  const handleSendMessage = (newMessage: string) => {
```
Comment (brunocroh) [inline apps/web/app/room/[slug]/page.tsx]: Add a useCallback here.

File: apps/web/components/chat.tsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+import { MessageCircle } from 'lucide-react'
+import { Message } from '@/app/room/[slug]/page'
+import { Button } from './ui/button'
+import { Input } from './ui/input'
+import { ScrollArea } from './ui/scroll-area'
+
+interface ChatProps {
+  messages: Message[]
+  onSend: (message: string) => void
+}
```
Comment (brunocroh) [inline apps/web/components/chat.tsx]: Replace this interface for a type

```suggestion
type ChatProps =  {
  messages: Message[]w
  onSend: (message: string) => void
}
```

Comment (brunocroh) [reply apps/web/components/chat.tsx]: https://www.totaltypescript.com/type-vs-interface-which-should-you-use

**Review by brunocroh — CHANGES_REQUESTED** [review]

**Comment by joaoreider** [comment]
Hey @brunocroh, i applied your suggestions. If there's anything else to do, just let me know

**Review by brunocroh — APPROVED** [review]

**Pull request merged**